src/parse_literals.cpp

```diff
@@ -1,25 +1,60 @@
 #include "MiniLua/values.hpp"
 
+#include <iterator>
 #include <regex>
 #include <sstream>
+#include <stdexcept>
 
 namespace minilua {
 
-// helper functions
+static std::string pattern_decimal = R"((\s*-?\d+\.?\d*))";
+static std::string pattern_hex = R"((\s*-?0[xX][\dA-Fa-f]+\.?[\dA-Fa-f]*))";
+static std::string pattern_scientific_notation = R"((\s*-?\d+\.?\d*[eE]-?\d+))";
+static std::regex to_number_general_pattern(
+    pattern_decimal + "|" + pattern_hex + "|" + pattern_scientific_notation,
+    std::regex::ECMAScript | std::regex::optimize);
+static std::regex to_number_int_pattern(R"(\s*-?([0-9]+|(0[xX][a-fA-F0-9]+)))");
 
 auto parse_number_literal(const std::string& str) -> Value {
-    std::regex pattern_number(R"(\s*\d+\.?\d*)");
-    std::regex pattern_hex(R"(\s*0[xX][\dA-Fa-f]+\.?[\dA-Fa-f]*)");
-    std::regex pattern_exp(R"(\s*\d+\.?\d*[eE]-?\d+)");
+    // if the string matches the expected format we parse it otherwise return nil
+    if (std::regex_match(str, to_number_int_pattern)) {
+        // parse as int and check if whole input is consumed
+        try {
+            char* last_pos;
+            // we need to parse as unsigned long to retain the same behaviour as lua
```

Comment:
Yes but if you parse e.g. `0xfff...fff` it should be -1. If we parse it as a signed long it will just overflow (or be the max long, I don't remember). If we parse it as an unsigned long and the cast it works the same as in lua. And the parse functions for unsigned longs also accepts `-something`. I have to write a few more tests to confirm that this works in all cases (especially for literals that are out of range) but all current tests pass.

Reply:
As it turns out this will not parse correctly if the value is out of range. But I think none of the C/C++ functions will parse it correctly if it is out of range.